core/src/contrib/systems/net/WsServerSystem.ts

```diff
@@ -3,22 +3,33 @@ import { System, TickData, World } from "@piggo-gg/core";
 export type ServerNetcodeSystemProps = {
   world: World
   clients: Record<string, { send: (_: string) => number }>
-  clientMessages: Record<string, { td: TickData, latency: number }>
+  latestClientMessages: Record<string, { td: TickData, latency: number }>
 }
 
-export const WsServerSystem = ({ world, clients, clientMessages }: ServerNetcodeSystemProps): System => {
+export const WsServerSystem = ({ world, clients, latestClientMessages }: ServerNetcodeSystemProps): System => {
 
   const onTick = () => {
 
     // prepare actions & messages for this tick + future ticks
-    const frames = world.actionBuffer.keys().filter((tick) => tick >= world.tick);
+    const frames = world.actionBuffer.keys().filter((tick) => tick >= world.tick).reverse();
+
     let actions: Record<number, Record<string, string[]>> = {};
     let chats: Record<number, Record<string, string[]>> = {};
 
+    // populate the first 15 ticks empty
+    for (let i = 0; i < 15; i++) actions[world.tick + i] = {}
+
+    const latestActionForEntity: Record<string, [number, string[]]> = {};
+
     frames.forEach((tick) => {
       const actionsAtTick = world.actionBuffer.atTick(tick);
       if (actionsAtTick && Object.keys(actionsAtTick).length) {
         actions[tick] = actionsAtTick;
+        Object.keys(actionsAtTick).forEach((playerId) => {
+          if (playerId.startsWith("skelly")) {
+            latestActionForEntity[playerId] = [tick, actionsAtTick[playerId]];
+          }
+        });
```

Comment:
this client prediction implementation needs refactored

whether an action should be "predicted" forward should be configurable. and it should be applied to all controlled entities generically